src/lang/defs/finder.rs

```diff
@@ -85,6 +88,87 @@ fn get_declaration_of(db: &AnalysisDatabase, def: &ResolvedItem) -> Option<Resol
     try_impl_items(db, &terminal_identifier)
 }
 
+fn try_trait_or_impl_item_with_self_reference(
+    db: &AnalysisDatabase,
+    identifier: &TerminalIdentifier,
+    lookup_items: &[LookupItemId],
+) -> Option<ResolvedItem> {
+    // Find if we're on ExprPath
+    let path_item_segments_origin =
+        identifier.as_syntax_node().ancestor_of_type::<ast::ExprPath>(db)?.to_segments(db);
+
+    // Snip off the path after the identifier we're on
+    let path_item_segments: Vec<_> = path_item_segments_origin
+        .iter()
+        .take_while_inclusive(|segment| segment.identifier(db) != identifier.text(db))
+        .cloned()
+        .collect();
+
+    let (first_segment, second_segment) = match path_item_segments.as_slice() {
+        [first, second, ..] => (first, Some(second)),
+        [first, ..] => (first, None),
+        _ => return None,
```

Comment:
```suggestion
        [first] => (first, None),
        [] => return None,
```
This is more explicit